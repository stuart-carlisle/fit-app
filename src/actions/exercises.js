import { database } from '../firebase/firebase'

//SET_EXERCISES_PUBLIC
export const setExercisesPublic = (exercisesPublic) => ({
    type:'SET_EXERCISES_PUBLIC',
    exercisesPublic
  })


//SET_EXERCISES_PRIVATE
export const setExercisesPrivate = (exercisesPrivate) => ({
  type:'SET_EXERCISES_PRIVATE',
  exercisesPrivate
})

export const startSetExercisesPrivate = () => {
  return (dispatch,getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/exercises`).once('value').then((snapshot)=>{
        const exercisesPrivate = []
        snapshot.forEach((childSnapshot)=>{
          exercisesPrivate.push({
            id:childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        dispatch(setExercisesPrivate(exercisesPrivate))
    })  
  }
}

//SET_EXERCISES_DIARY
export const setExercisesDiary = (exercises) => ({
  type:'SET_EXERCISES_DIARY',
  exercises
})

export const startSetExercisesDiary = () => {
  return (dispatch,getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/diary/exercises`).once('value').then((snapshot)=>{
        const exercises = []
        snapshot.forEach((childSnapshot)=>{
          exercises.push({
            id:childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        dispatch(setExercisesDiary(exercises))
    })  
  }
}

// ADD_EXERCISE PRIVATE
export const addExercisePrivate = (exercise) => ({
  type: 'ADD_EXERCISE_PRIVATE',
  exercise
});

export const startAddExercisePrivate = (exerciseData) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const { description, type, energy, time, lastUsed } = exerciseData
    const exercise = { description, type, energy, time, lastUsed }
   
    return database.ref(`users/${uid}/exercises`).push(exercise).then((ref)=>{
      dispatch(addExercisePrivate({
        id:ref.key,
        ...exercise
      }))
    })
  }
}

// ADD_EXERCISE_DIARY
export const addExerciseDiary = (exercise) => ({
  type: 'ADD_EXERCISE_DIARY',
  exercise
});

export const startAddExerciseDiary = (exerciseData) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const { description, type, energy, time, diaryDate } = exerciseData
    const exercise = { description, type, energy, time, diaryDate }
   
    return database.ref(`users/${uid}/diary/exercises`).push(exercise).then((ref)=>{
      dispatch(addExerciseDiary({
        id: ref.key,
        ...exercise
      }))
    })
  }
}
        
// UPDATE_EXERCISE_PRIVATE
export const updateExercisePrivate = (id, updates) => ({
  type: 'UPDATE_EXERCISE_PRIVATE',
  id,
  updates
});

export const startUpdateExercisePrivate = (id, updates) => {
  return (dispatch,getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/exercises/${id}`).update(updates).then(()=>{
      dispatch(updateExercisePrivate(id,updates))
    })
  }
}

// REMOVE_EXERCISE_DIARY
export const removeExerciseDiary = ({ id } = {}) => ({
  type: 'REMOVE_EXERCISE_DIARY',
  id
});


export const startRemoveExerciseDiary = ({ id } = {}) => {
  return (dispatch,getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/diary/exercises/${id}`).remove().then(()=>{
      dispatch(removeExerciseDiary({id}))
    })
  }
}

// UPDATE_EXERCISE_DIARY
export const updateExerciseDiary = (id, updates) => ({
  type: 'UPDATE_EXERCISE_DIARY',
  id,
  updates
});

export const startUpdateExerciseDiary = (id, updates) => {
  return (dispatch,getState) => {
  const uid = getState().auth.uid
  return database.ref(`users/${uid}/diary/exercises/${id}`).update(updates).then(()=>{
    dispatch(updateExerciseDiary(id,updates))
  })
}}