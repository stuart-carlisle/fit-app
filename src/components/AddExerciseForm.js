import React, { useState , useEffect} from 'react'
import moment from 'moment'
import { connect } from 'react-redux'
import numeral from 'numeral'

export const AddExerciseForm = ({
  filters = {},
  onSubmit, 
  exercise = {
    description:'',
    type:'',
    energy:'',
    time:''
  }
}) => {
    const defaultEnergy = numeral(exercise.energy).value()
    const defaultTime = numeral(exercise.time).value()
    const date = filters.date
    const referenceDatabase = filters.referenceDatabase
    const [time, setTime] = useState(defaultTime)
    const [error, setError] = useState('')
    const [energy, setEnergy] = useState(defaultEnergy)
    const diaryDate = date ? date.valueOf() : moment().startOf('day').valueOf()
    const lastUsed = moment().valueOf()

    useEffect(()=>{
        
      let refreshed = JSON.parse(localStorage.getItem('refreshed'))
      if(refreshed){
        window.history.back()
        window.scrollTo(0,0)
        localStorage.removeItem('refreshed')
      }else{
        localStorage.setItem('refreshed',JSON.stringify(true))
      }
    },[])

    const onFormSubmit = (e) => {
        e.preventDefault();
        const exerciseData = { 
          description: exercise.description,
          type: exercise.type
        }
        if (numeral(time).value()<=0) {
          setError('Please provide a time greater than 0')
          window.scrollTo(0,0)
        } else {
          setError('')
          onSubmit(
            referenceDatabase,
            diaryDate,
            lastUsed,
            exerciseData,
            time.toString(),
            energy.toString()
          )
          localStorage.removeItem('refreshed')
        }
      }
    
    const onTimeChange = (e)=>{
        const time = e.target.value
          if (!time || time.match(/^\d{1,}(\.\d{0,1})?$/)) {
          setTime(time)
          const timeValue = numeral(time).value()
          setEnergy(Math.round(defaultEnergy * (timeValue/defaultTime)))
        }
    }

    const onEnergyChange = ()=>{
      
    }

    return (
        <div>
        <div className="content-container__background">
        <div className="content-container__form__update">
            <form className= "form" onSubmit={onFormSubmit}>
                {error && <p id="error-message" className="form__error">{error}</p>}
                  <h1 className="page-header__title-update"> {exercise.description} </h1>
                    <span className="text-input__span-container-form">
                      <div className="content-container__update-titles">
                        <h3 className="form__heading__update">Exercise Time (mins):</h3>
                        <h3 className="form__heading__update">Energy (calories): </h3>
                      </div>
                      <div className="content-container__update">
                        <input  //Exercise Duration
                          className="text-input__update-form"
                          type="text"
                          value={time}
                          onChange={onTimeChange}
                          placeholder="Enter a Duration"
                        />  
                        <input  //Energy Used
                          className="text-input__update-form"
                          type="text"
                          value={energy?energy:''}
                          onChange={onEnergyChange}
                          placeholder="Energy Used"
                        /> 
                      </div>
                    </span>
                <button className="button button--form">Add Exercise</button>
            </form>
          </div>
        </div>
      </div>  
    )
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}


export default connect(mapStateToProps)(AddExerciseForm);