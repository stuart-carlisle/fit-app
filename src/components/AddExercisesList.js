import React from 'react'
import { connect } from 'react-redux'
import AddExercisesListItem from './AddExercisesListItem'
import selectExercises from '../selectors/exercises'

export const AddExercisesList = ({exercises}) => {
  let exercisesTopResults
  
  if(exercises.length>10){
    exercisesTopResults = exercises.slice(0,10)
  }else{
    exercisesTopResults = exercises
  }

  return(<div className="content-container">
   <div className="list-header">
    <div className="">Exercise</div>
    <div className="show-for-tablet">Calories Burned</div>
    <div className="show-for-mobile">Cals Burned</div>
   </div>
    <div className="list-body">
      {
        exercisesTopResults.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No Exercises In History</span>
          </div>
        ) : (
         exercisesTopResults.map((exercise) => {
            return <AddExercisesListItem key={exercise.id} {...exercise} />
          })
        )
      }
    </div>
  </div>  
  )
}

const mapStateToProps = (state) => {
  return {
    exercises: state.filters.referenceDatabase === 'public' ? state.exercisesPublic : selectExercises(state.exercisesPrivate,  state.filters)
  }
}

export default connect(mapStateToProps)(AddExercisesList);
