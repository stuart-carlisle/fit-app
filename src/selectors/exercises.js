// Get visible expenses

export default (exercises, { text, sortBy }) => {
  
  return exercises.filter((exercise) => {
      const textMatch = exercise.description.toLowerCase().includes(text.toLowerCase())
      return textMatch;
    }).sort((a, b) => {
      if (sortBy === 'date') {
        return a.lastUsed < b.lastUsed ? 1 : -1;
      } else if (sortBy === 'alphabetical') {
        return a.description > b.description ? 1 : -1;
      }
    })
  } 
