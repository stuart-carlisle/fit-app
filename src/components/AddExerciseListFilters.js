import React, { useState } from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAlphabetical, setReferenceDatabaseToPublic } from '../actions/filters';
import { setExercisesPublic } from '../actions/exercises'
import algoliasearch from 'algoliasearch/lite'

export const AddExerciseListFilters = ({
    filters,
    setTextFilter, 
    setReferenceDatabasePublic,
    sortByDate, 
    sortByAlphabetical,
    setExercisesPublic 
  }) => {
 
  const searchClient = algoliasearch (
      process.env.ALGOLIA_APP_ID,
      process.env.ALGOLIA_API_KEY
  )

  const index = searchClient.initIndex('exercises');

  const [text, setText] = useState('')
    
  const onTextChange = (e) => {
    setTextFilter(e.target.value)
    setText(e.target.value)
  }

  const onSortChange = (e) => {
    if (e.target.value === 'date') {
      sortByDate()
    } else if (e.target.value === 'alphabetical') {
      sortByAlphabetical()
    }
  }

  const onSubmit = (e) => {
    e.preventDefault()
    index
      .search({
        query: text
      })
      .then((responses) => {
        const array = responses.hits
        const newExercises = array.map((exercise)=>{
          return{
            description:exercise.description,
            type:exercise.type,
            energy:exercise.energy,
            time:exercise.time,
            id:exercise.objectID
          }
        })
        setExercisesPublic(newExercises)
      }).then(()=>{
        setReferenceDatabasePublic()
      })
}

    return (
      
      <div className="header__content--filters"> 
      <form className="filters-text-input__span-container" onSubmit={onSubmit}>
          <div className="input-group__item-page-header">
            <input
              type="text"
              className="filters-text-input"
              placeholder="SEARCH EXERCISES"
              value={text}
              onChange={onTextChange}
            />
          </div>
          <div className="input-group__item-page-header">
            <select
              className="minimal"
              value={filters.sortBy}
              onChange={onSortChange}
            >
              <option value="date">DATE</option>
              <option value="alphabetical">ALPHABETICAL</option>
            </select>
          </div>
            <button className="button filters-button--form-page-header">Search Online</button>
            </form>   
        </div>
    )
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAlphabetical: () => dispatch(sortByAlphabetical()),
  setReferenceDatabasePublic: () => dispatch(setReferenceDatabaseToPublic()),
  setExercisesPublic: (exercises) => dispatch(setExercisesPublic(exercises))
})

export default connect(mapStateToProps,mapDispatchToProps)(AddExerciseListFilters);

