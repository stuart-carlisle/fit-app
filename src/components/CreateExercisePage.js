import React from 'react'
import { connect } from 'react-redux';
import ExerciseForm from './ExerciseForm';
import { startAddExercisePrivate, startAddExerciseDiary } from '../actions/exercises';
import { Link } from 'react-router-dom'
import { databasePublic } from '../firebase/firebase'

export const CreateExercisePage = (props) => {
    
    const onSubmit = (exercise) => {
      const privateExercise = { 
        ...exercise,
        description: exercise.description.toUpperCase()
      }
      const publicExercise = { 
        description: exercise.description.toUpperCase(), 
        type: exercise.type, 
        energy: exercise.energy,
        time: exercise.time 
      }
        databasePublic.ref('exercises').push(publicExercise)
        props.startAddExercisePrivate(privateExercise)
        props.startAddExerciseDiary(privateExercise)
        props.history.push('/diary');
    }
     
    return(
        <div>
        <Link className="button__back-button" to={'/add-exercise'}><img className="image__header" src="/images/back-arrow.svg"/></Link>
        <div className="page-header__filter page-header__filter--with-back">
            <div className="content-container">
              <h1 className="page-header__title">CREATE EXERCISE</h1>
            </div>
          </div>
          <div className="show-for-mobile page-header__margin"></div>
          <div className="content-container">
            <ExerciseForm
              id="exercise-form"
              onSubmit={onSubmit}
            />
          </div>
        </div>
    )
}
  
const mapDispatchToProps = (dispatch) =>({
    startAddExercisePrivate: (exercise) => dispatch(startAddExercisePrivate(exercise)),
    startAddExerciseDiary: (exercise) => dispatch(startAddExerciseDiary(exercise))
})
  
export default connect(undefined, mapDispatchToProps)(CreateExercisePage);
