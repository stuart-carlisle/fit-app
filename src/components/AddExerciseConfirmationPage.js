import React from 'react'
import { connect } from 'react-redux'
import AddExerciseForm from './AddExerciseForm'
import { Link } from 'react-router-dom'
import { startAddExerciseDiary, startAddExercisePrivate, startUpdateExercisePrivate } from '../actions/exercises'

export const AddExerciseConfirmationPage = (props) => {
  
  const onSubmit = (referenceDatabase, diaryDate, lastUsed, exerciseDescAndType, time, energy) => {
    if (referenceDatabase==='public'){
      props.startAddExercisePrivate({...exerciseDescAndType, time, energy, lastUsed})
    } else if (referenceDatabase==='private'){
      props.startUpdateExercisePrivate(props.exercise.id,{...exerciseDescAndType, time, energy, lastUsed})
    }
    props.startAddExerciseDiary({...exerciseDescAndType, time, diaryDate, energy})
    props.history.push('/diary');
  }
  

  return (
     <div>
     <Link className="button__back-button" to={`/add-exercise`}><img className="image__header" src="/images/back-arrow.svg"/></Link>
        <div className="page-header__filter page-header__filter--with-back">
          <div className="content-container">
            <h1 className="page-header__title">ADD EXERCISE</h1>
          </div>
        </div>
        <div className="show-for-mobile page-header__margin"></div>
        <div className="content-container">
        <AddExerciseForm id="add-exercise-form" exercise={props.exercise} onSubmit={onSubmit} />
        </div>
     </div>
 )
}

const mapStateToProps = (state,props) => {
  if(state.filters.referenceDatabase==='public'){
    return {
      exercise: state.exercisesPublic.find((exercise) => exercise.id === props.match.params.id)
    }
  }else if(state.filters.referenceDatabase==='private'){
    return {
      exercise: state.exercisesPrivate.find((exercise) => exercise.id === props.match.params.id)
    }
  }
}

const mapDispatchToProps = (dispatch) =>({
    startAddExerciseDiary: (exercise) => dispatch(startAddExerciseDiary(exercise)),
    startUpdateExercisePrivate: (id, exercise) => dispatch(startUpdateExercisePrivate(id, exercise)),
    startAddExercisePrivate: (exercise) => dispatch(startAddExercisePrivate(exercise))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddExerciseConfirmationPage);




