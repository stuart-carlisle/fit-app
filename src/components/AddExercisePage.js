import React, { useEffect } from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import AddExerciseListFilters from './AddExerciseListFilters'
import AddExercisesList from './AddExercisesList'
import { setTextFilter, sortByDate, setReferenceDatabaseToPrivate } from '../actions/filters'

export const AddExercisePage = ({ setTextFilter, sortByDate, setReferenceDatabaseToPrivate}) => {
  useEffect(()=>{
    localStorage.removeItem('refreshed')
      setTextFilter('')
      sortByDate()
      setReferenceDatabaseToPrivate()
  },[])
    return(
      <div>
      <Link className="button__back-button" to={`/diary`}><img className="image__header" src="/images/back-arrow.svg"/></Link>
    <div className="page-header__filter page-header__filter--with-back">
      <div className="content-container">
        <AddExerciseListFilters />
      </div>  
    </div>
    <AddExercisesList />
    <div className="content-container">
    <Link className="button button--link" to={`/create-exercise`}>+ CREATE CUSTOM EXERCISE</Link>
    </div>
    </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  setReferenceDatabaseToPrivate: () => dispatch(setReferenceDatabaseToPrivate())
})

export default connect (undefined, mapDispatchToProps)(AddExercisePage)