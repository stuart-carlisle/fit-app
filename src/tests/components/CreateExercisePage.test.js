import React from 'react'
import { shallow } from 'enzyme'
import { CreateExercisePage   } from '../../components/CreateExercisePage'
import exercises from '../fixtures/exercises'

test('Should render CreateExercisePage with props correctly',()=>{
    const startAddExercisePrivate = jest.fn()
    const startAddExerciseDiary = jest.fn()
    const history = { push: jest.fn() }
    const wrapper = shallow(
        <CreateExercisePage 
            startAddExerciseDiary={startAddExerciseDiary}
            startAddExercisePrivate={startAddExercisePrivate} 
            history={history}
        />
    )
    expect(wrapper).toMatchSnapshot()
})

test('should call the startAddExercisePrivate and startAddExerciseDiary actions when submitted',()=>{
    const {id, ...exercise} = exercises[0]
    const privateExercise = {
        ...exercise,
        description: exercise.description.toUpperCase()
    }
    const startAddExercisePrivate = jest.fn()
    const startAddExerciseDiary = jest.fn()
    const history = { push: jest.fn() }
    const wrapper = shallow(
        <CreateExercisePage 
            startAddExerciseDiary={startAddExerciseDiary}
            startAddExercisePrivate={startAddExercisePrivate} 
            history={history}
        />
    )
    wrapper.find('#exercise-form').prop('onSubmit')(privateExercise)
    expect(startAddExercisePrivate).toHaveBeenLastCalledWith(privateExercise)
    expect(startAddExerciseDiary).toHaveBeenLastCalledWith(privateExercise)
    expect(history.push).toHaveBeenCalledWith('/diary')
})



