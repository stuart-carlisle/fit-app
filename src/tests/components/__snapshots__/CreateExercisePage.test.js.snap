// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render CreateExercisePage with props correctly 1`] = `
<div>
  <Link
    className="button__back-button"
    replace={false}
    to="/add-exercise"
  >
    <img
      className="image__header"
      src="/images/back-arrow.svg"
    />
  </Link>
  <div
    className="page-header__filter page-header__filter--with-back"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        CREATE EXERCISE
      </h1>
    </div>
  </div>
  <div
    className="show-for-mobile page-header__margin"
  />
  <div
    className="content-container"
  >
    <Connect(ExerciseForm)
      id="exercise-form"
      onSubmit={[Function]}
    />
  </div>
</div>
`;
